default/python/AI/colonization/calculate_research.py

```diff
@@ -0,0 +1,213 @@
+    skill_multiplier = get_species_tag_value(species.name, AIDependencies.Tags.RESEARCH)
+    bonus_by_policy = _get_research_bonus_modified_by_policy(stability)
+    flat_by_policy = _get_research_flat_modified_by_policy(planet, stability)
+    policy_multiplier = _get_policy_multiplier(stability)
+    bonus_unmodified = _get_research_bonus_unmodified(planet, stability)
+
+    result = max_population * (AIDependencies.INDUSTRY_PER_POP + bonus_modified) * skill_multiplier
+    result = (result + max_population * bonus_by_policy + flat_by_policy) * policy_multiplier
+    result += max_population * bonus_unmodified
+    if DEBUG_COLONY_RATING:
+        debug(
+            f"calculate_research pop={max_population:.2f}, st={stability:.2f}, b1={bonus_modified:.2f}, "
+            f"m1={skill_multiplier:.2f}, b2={bonus_by_policy:.2f}, f2={flat_by_policy:.2f}, "
+            f"m2={policy_multiplier:.2f}, b3={bonus_unmodified:.2f} -> {result:.2f}"
+        )
+    return result
+
+
+class _ResearchBonus(NamedTuple):
+    available: bool
+    min_stability: float
+    value: float
+
+
+@cache_for_current_turn
+def _get_modified_research_bonuses(planet: fo.planet) -> List[_ResearchBonus]:
+    """
+    Get list of per-population bonuses which are added before multiplication with the species skill value.
+    """
+    # use fo.getEmpire().researchQueue directly here to simplify caching this function
+    specials = planet.specials
+    return [
+        _ResearchBonus(
+            AIDependencies.ANCIENT_RUINS_SPECIAL in specials or AIDependencies.ANCIENT_RUINS_SPECIAL2 in specials,
+            get_named_real("ANCIENT_RUINS_MIN_STABILITY"),
+            get_named_real("ANCIENT_RUINS_TARGET_RESEARCH_PERPOP"),
+        ),
+        _ResearchBonus(
+            fo.getEmpire().policyAdopted("PLC_DIVERSITY"),
+            get_named_real("PLC_DIVERSITY_MIN_STABILITY"),
+            (len(get_empire_planets_by_species()) - get_named_int("PLC_DIVERSITY_THRESHOLD"))
+            * get_named_real("PLC_DIVERSITY_SCALING"),
+        ),
+        _ResearchBonus(
+            tech_soon_available("GRO_ENERGY_META", 3),
+            get_named_real("GRO_ENERGY_META_MIN_STABILITY"),
+            get_named_real("GRO_ENERGY_META_TARGET_RESEARCH_PERPOP"),
+        ),
+        _ResearchBonus(
+            bool(BuildingType.ENCLAVE_VOID.built_or_queued_at()),
+            get_named_real("BLD_ENCLAVE_VOID_MIN_STABILITY"),
+            get_named_real("BLD_ENCLAVE_VOID_TARGET_RESEARCH_PERPOP"),
+        ),
+    ]
+
+
+def _get_research_bonus_modified(planet: fo.planet, stability: float) -> float:
+    """
+    Calculate bonus research per population which would be added before multiplication with the species skill value.
+    """
+    result = 0.0
+    for bonus in _get_modified_research_bonuses(planet):
+        if bonus.available and stability >= bonus.min_stability:
+            result += bonus.value
+    return result
```

Comment:
There is an interesting approach to designing objects responsible for data. 
Instead of doing logic outside the object, we can do the logic inside it. 

Making _ResearchBonus to work in that way  will make this code could looks like that:
```suggestion
    return sum(bonus.get_bonus(stability) for for bonus in _get_modified_research_bonuses(planet))
```

Reply:
That's definitely better